consumer/consumererror/README.md

```diff
@@ -0,0 +1,300 @@
+code returned from a backend in a response to a system upstream from the
+Collector.
+
+## Current targets for using errors
+
+**Receivers**: Receivers should be able to consume multiple errors downstream
+and determine the best course of action based on the user's configuration. This
+may entail either keeping the retry queue inside of the Collector by having the
+receiver keep track of retries, or may involve having the caller manage the
+retry queue by returning a retryable network code with relevant information.
+
+**scraperhelper**: The scraper helper can use information about errors from
+downstream to affect scraping. For example, if the backend is struggling with
+the volume of data, scraping could be slowed, or the amount of data collected
+could be reduced. 
+
+**exporterhelper**: When an exporter returns a retryable error, the
+exporterhelper can use this information to retry. Permanent errors will be
+forwarded back up the pipeline.
+
+**obsreport**: Recording partial success information can ensure we correctly
+track the number of failed telemetry records in the pipeline. Right now, all
+records will be considered to be failed, which isn't accurate when partial
+successes occur.
+
+## Creating Errors
+
+Errors can be created by calling `consumererror.New(err, opts...)` where `err`
+is the underlying error, and `opts` is one of the provided options for supplying
+additional metadata:
+
+- `consumererror.WithGRPCStatus`
+- `consumererror.WithHTTPStatus`
+
+The following options are not currently available, but may be made available in
+the future:
+
+- `consumererror.WithRetry`
+- `consumererror.WithPartial`
+- `consumererror.WithMetadata`
+
+All options can be combined, we assume that the component knows what it is doing
+when seemingly conflicting options.
+
+Two examples:
+
+- `WithRetry` and `WithPartial` are included together: Partial successes are
+  considered permanent errors in OTLP, which conflicts with making an error
+  retryable by including `WithRetry`. However, per our definition of what makes
+  a permanent error, this error has been marked as retryable, and therefore we
+  assume the component producing this error supports retyable partial success
+  errors.
+- `WithGRPCStatus` and `WithHTTPStatus` are included together: While the
+  component likely only sent data over one of these transports, our errors will
+  produce the given status if it is included on the error, otherwise it will
+  translate a status from the status for the other transport. If both of these
+  are given, we assume the component wanted to perform its own status
+  conversion, and we will simply give the status for the requested transport
+  without performing any conversion.
+
+**Example**:
+
+```go
+consumererror.New(err,
+  consumererror.WithRetry(
+    consumerrerror.WithRetryDelay(10 * time.Second)
+  ),
+  consumererror.WithGRPCStatus(codes.InvalidArgument),
+)
```

Comment:
This example must be invalid. `codes.InvalidArgument` should not be retryable. 

In general, it still seems like all these options provide too much flexibility, making it confusing to use and easy to misuse. 

For example, why GRPC and HTTP have to be options? Why can't we just have different constructors like `consumererror.NewFromGRPC(grpc.Status)`? The status should have the retry information that we can retrieve if applicable. If we expect exporters to add extra info on top of grpc error, they can manually parse the grpc status and create a collector "internal" error another way.